mlir/lib/Dialect/Rock/Transforms/LowerRockReduce.cpp

```diff
@@ -0,0 +1,233 @@
+//===- LowerRockReduce.cpp - The lowering pass of rock.reduce ------------===//
+//
+// Copyright 2022 Advanced Micro Devices.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================
+//
+// This pass converts rock.reduce into the TransformingFor loop
+// using global atomics.
+//
+//===-----------------------------------------------------===//
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/IR/TransformMapBuilder.h"
+#include "mlir/Dialect/Rock/Passes.h"
+#include "mlir/Dialect/Rock/utility/transformMapUtils.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Support/LogicalResult.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "llvm/Support/Debug.h"
+#include <memory>
+
+namespace mlir {
+namespace rock {
+#define GEN_PASS_DEF_ROCKLOWERREDUCEPASS
+#include "mlir/Dialect/Rock/Passes.h.inc"
+} // namespace rock
+} // namespace mlir
+
+#define DEBUG_TYPE "rock-lower-rock-reduce"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+namespace {
+class RockLowerReducePass
+    : public rock::impl::RockLowerReducePassBase<RockLowerReducePass> {
+  void runOnOperation() override;
+};
+
+struct ReduceRewritePattern : public OpConversionPattern<ReduceOp> {
+  using OpConversionPattern<ReduceOp>::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(ReduceOp op, ReduceOpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override;
+};
+} // end namespace
+
+// This function will create a view of the tensor that is being
+// reduced from arbitary dimension sizes to bid x tid x iter space.
+static Value createThreadView(Value redInput, int64_t blockSize,
+                              int64_t gridSize, Location loc,
+                              PatternRewriter &rewriter) {
+  auto totalThreads = gridSize * blockSize;
+  auto inpShape = redInput.getType().cast<ShapedType>();
+  auto elementCount = inpShape.getNumElements();
+  auto dataPerThread = (elementCount + (totalThreads - 1)) / totalThreads;
+
+  llvm::errs() << "totalThreads = " << totalThreads << "\n";
```

Comment:
`LLVM_DEBUG(llvm::dbgs() <<`

Reply:
same here -- consequence me commiting to dev branch and moving servers. I m planning to remove these. Apologies about the noise.